File: JS 문제 풀이/프로그래머스 Lv.1/크레인 인형뽑기 게임.js
function solution(board, moves) {
    let answer = 0;
    let stack = [];
    const N=board.length;//격자 크기

    for(let i=0;i<moves.length;i++){
        let col = moves[i]-1;
        let doll = 0;

        for(let j=0;j<N;j++){
            if(board[j][col] !=0 ){// 인형이 존재할 경우
                doll=board[j][col]; // 1. 인형을 집는다
                board[j][col]=0; // 2. 해당 격자를 비운다.
                break;
            }
        }
        if(doll !=0 ){ // 인형을 집은 경우
            if(stack.length == 0 ) stack.push(doll); // stack에 인형이 없는 경우
            else{ // stack에 인형이 있는 경우
                if(doll == stack[stack.length-1]){ // stack의 마지막 인형이 같은 인형일 때
                    stack.pop();
                    answer+=2;
                }
                else { // stack의 인형이 같은 인형이 아닐 때
                    stack.push(doll);
                }
            }
        }else{ // 인형을 못 집은 경우
            continue;
        }
    }
    return answer;
}

console.log(solution([[0, 0, 0, 0, 0], [0, 0, 1, 0, 3], [0, 2, 5, 0, 1], [4, 2, 4, 4, 2], [3, 5, 1, 3, 1]], [1, 5, 3, 5, 1, 2, 1, 4])); // 4

/*
sol. 인형 뽑기 게임으로 간단하게 stack을 이용한 구현 문제이다.
*/

File: JAVA 문제풀이/프로그래머스 Lv.1/크레인 인형뽑기 게임.java
/*
Java의 Stack을 사용해보자
*/
import java.util.Stack;
class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        Stack<Integer> stack = new Stack<>();

        for(int value:moves){
            int col = value-1;
            int doll=0;

            for(int i=0;i<board.length;i++){
                if(board[i][col]!=0){
                    doll=board[i][col];
                    board[i][col]=0;
                    break;
                }
            }

            if(doll!=0){
                if(stack.size()==0){
                    stack.push(doll);
                }
                else{
                    if(stack.peek()==doll){
                        stack.pop();
                        answer++;
                    }
                    else{
                        stack.push(doll);
                    }
                }
            }
        }

        return answer*2;
    }
}
